static/js/pill_typeahead.js

```diff
@@ -79,19 +86,29 @@ export function set_up(input, pills, opts) {
                 return item.name.toLowerCase().includes(query);
             }
 
+            let matches = false;
+
             if (include_user_groups) {
-                return group_matcher(query, item) || person_matcher(query, item);
+                matches = matches || group_matcher(query, item);
             }
 
-            return person_matcher(query, item);
+            if (include_users) {
+                matches = matches || person_matcher(query, item);
+            }
+
+            return matches;
```

Comment:
I'm not sure this structure makes sense; if we're trying to match a particular item, could we just do:
```
if (include_user_groups && is_user_group(item)) {
    return group_matcher(query, item);
}
```

since we should know what type `item` is?  

(I don't know if we actually have that detail plumbed and available here, but that feels like a more natural way to frame the question)

Reply:
We have details, for writing it as you suggested. But `group_matcher` alraedy does that check, before trying to match the item, and returns `false` if item is not a user group. So, writing it as you suggested will cause us having two checks for same thing one after another.
Current implementation also appears similar to as you suggested but we just check for whether item is a group or not inside `group_matcher` function, this also makes `group_matcher` capable of handling items which are not group .
I haven't changed it for now but if you still feel that it should be, then please mention I would update that.